core/src/main/java/com/linecorp/armeria/server/ServiceRequestContextBuilder.java

```diff
@@ -228,26 +227,32 @@ public ServiceRequestContext build() {
         final InetAddress clientAddress = server.config().clientAddressMapper().apply(proxiedAddresses)
                                                 .getAddress();
 
-        final CancellationScheduler requestCancellationScheduler;
+        final DefaultServiceRequestContext ctx;
+        // Build the context with the properties set by a user and the fake objects.
+        final Channel ch = fakeChannel();
         if (timedOut()) {
-            requestCancellationScheduler = CancellationScheduler.finished(true);
+            ctx = new DefaultServiceRequestContext(
+                    serviceCfg, ch, eventLoop(), meterRegistry(), sessionProtocol(), id(), routingCtx,
+                    routingResult, exchangeType, req, sslSession(), proxiedAddresses,
+                    clientAddress, remoteAddress(), localAddress(),
+                    CancellationScheduler.finished(true),
+                    isRequestStartTimeSet() ? requestStartTimeNanos() : System.nanoTime(),
+                    isRequestStartTimeSet() ? requestStartTimeMicros() : SystemInfo.currentTimeMicros(),
+                    HttpHeaders.of(), HttpHeaders.of(), serviceCfg.contextHook());
         } else {
-            requestCancellationScheduler = CancellationScheduler.ofServer(0);
-            requestCancellationScheduler.initAndStart(eventLoop(), noopCancellationTask);
+            ctx = new DefaultServiceRequestContext(
+                    serviceCfg, ch, eventLoop(), meterRegistry(), sessionProtocol(), id(), routingCtx,
+                    routingResult, exchangeType, req, sslSession(), proxiedAddresses,
+                    clientAddress, remoteAddress(), localAddress(),
+                    CancellationScheduler.ofServer(0),
```

Comment:
- Should we extract `CancellationScheduler.ofServer(0)` as a variable instead of having two branches for `DefaultServiceRequestContext`.
- EventLoop should not be not-null when created.

```java
EventLoop eventLoop = eventLoop();
if (eventLoop == null) {
    eventLoop = CommonPools.workerGroup().next();
}

if (timedOut()) {
  cancellationScheduler = CancellationScheduler.finished(true);
} else {
  cancellationScheduler = cancellationScheduler.ofServer(0)
}
DefaultServiceRequestContext ctx = new DefaultServiceRequestContext(...);
```

Reply:
Thank you for your sincere review. 🙇  I will do my best to fix it!